### PR TITLE
fix issues around limiting, fix recursive extractStack

### PR DESCRIPTION
**Pull request by agroszer**

### CODE REVIEW

File: src/zope/exceptions/exceptionformatter.py [hunk 1]
```diff
@@ -200,16 +204,37 @@ def extractStack(self, f=None):
...
+            '...%(omitted)d entries omitted, because limit is %(limit)d...\n')
         result.reverse()
         return result
 
+    def _obeyLimit(self, result, template):
+        limit = self.getLimit()
+        if limit is not None and len(result) > limit:
+            # cut out the middle part of the TB
+            tocut = len(result) - limit
+            middle = len(result) / 2
```
Comment (tseaver) [inline src/zope/exceptions/exceptionformatter.py]: Use `//` for Py3k compatibility.

File: src/zope/exceptions/exceptionformatter.py [hunk 2]
```diff
@@ -200,16 +204,37 @@ def extractStack(self, f=None):
...
+            lower = middle - tocut / 2
```
Comment (tseaver) [inline src/zope/exceptions/exceptionformatter.py]: Ditto.

**Comment by agroszer** [comment]
thought so... gotta improve on 3.x

**Comment by mgedmin** [comment]
LGTM.

I'd mildly prefer `'blah blah {omitted} blah'.format(omitted=42)` to old-style `'blah blah %(omitted)d blah' % dict(omitted=42)`, but that's a minor.

I'm also somewhat curious about the extractStack() recursion protection.  It's to defend against bugs in the exception formatter itself?

**Comment by agroszer** [comment]
extractStack: 
holy-smoly, I added that in a9f60fb4a40a01c4d52bdac071c1a3ec43281be1, of course forgot why
test_extractStack_w_recursive_frames_and_limit showed that removing the limit sent extractStack into an endless loop, that code is just a hunch based on formatException

**Comment by tseaver** [comment]
@agroszer LGTM

**Pull request merged**